File: src/main/kotlin/dev/jefy/connectpro/shared/infrastructure/file_storage/AwsS3ImageServiceImpl.java
package dev.jefy.connectpro.shared.infrastructure.file_storage;

import org.jspecify.annotations.NullMarked;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

import dev.jefy.connectpro.shared.domain.vo.ImageUrl;
import lombok.RequiredArgsConstructor;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;

/**
 * @author Jôph Yamba
 */
@NullMarked
@Service
@RequiredArgsConstructor
class AwsS3ImageServiceImpl implements ImageService {

    private final ImageUtils imageUtils;


    private final S3Client s3Client;

    @Value("${aws.s3.bucket-name}")
    private String s3Bucket;


    @Override
    public ImageUrl save(MultipartFile imageFile) throws IOException {
        String extension = imageUtils.getExtension(imageFile);
        String imageName = imageUtils.createUniqueImageName(extension);
        PutObjectRequest objectRequest = PutObjectRequest
                .builder()
                .bucket(s3Bucket)
                .key(imageName)
                .contentType(imageFile.getContentType())
                .build();

        s3Client.putObject(
                objectRequest,
                RequestBody.fromBytes(imageFile.getBytes())
        );
        return new ImageUrl(imageUtils.fileUrl(imageName));
    }


    @Override
    public void delete(ImageUrl image) throws IOException {
        String imageUrl = image.value();
        String bucketUrlPrefix = imageUtils.getBucketUrlPrefix();
        if (!imageUrl.startsWith(bucketUrlPrefix)) {
            throw new IllegalArgumentException("Invalid file URL: " + imageUrl);
        }

        // Extract the S3 key from the URL
        String key = imageUrl.substring(bucketUrlPrefix.length());

        // Delete the object
        DeleteObjectRequest deleteObjectRequest = DeleteObjectRequest.builder()
                .bucket(s3Bucket)
                .key(key)
                .build();

        s3Client.deleteObject(deleteObjectRequest);
    }
}

File: src/main/kotlin/dev/jefy/connectpro/shared/infrastructure/file_storage/AwsS3ImageServiceImpl.kt
package dev.jefy.connectpro.shared.infrastructure.file_storage

import dev.jefy.connectpro.shared.domain.vo.Image
import dev.jefy.connectpro.shared.infrastructure.config.AwsProperties
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import software.amazon.awssdk.core.sync.RequestBody
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest
import software.amazon.awssdk.services.s3.model.PutObjectRequest
import java.io.IOException

@Service
class AwsS3ImageServiceImpl(
    private val imageUtils: ImageUtils,
    private val s3Client: S3Client,
    private val awsProperties: AwsProperties
) : ImageService {
    
    @Throws(IOException::class)
    override fun save(imageFile: MultipartFile): Image {
        val extension = imageUtils.getExtension(imageFile)
        val imageName = imageUtils.createUniqueImageName(extension)
        val objectRequest = PutObjectRequest.builder()
            .bucket(awsProperties.s3.bucketName)
            .key(imageName)
            .contentType(imageFile.contentType)
            .build()
        
        s3Client.putObject(objectRequest, RequestBody.fromBytes(imageFile.bytes))
        return Image(imageName)
    }

    @Throws(IOException::class)
    override fun delete(image: Image) {
        val deleteObjectRequest = DeleteObjectRequest.builder()
            .bucket(awsProperties.s3.bucketName)
            .key(image.value)
            .build()

        s3Client.deleteObject(deleteObjectRequest)
    }
}